Fix #7383: Add functionality to regenerate exploration opportunity summary models.

DubeySandeep opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #7383, adds functionality to regenerate exploration opportunity summary models.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

DubeySandeep committed:
AAded functionality to regenerate exp_opp_summary_models.

DubeySandeep committed:
Fixes remining issues in #7109.

DubeySandeep committed:
Undo backend test releated changes.

DubeySandeep commented on core/domain/opportunity_services.py:
```diff
@@ -384,3 +403,32 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+    story_ids = []
+    topics = topic_fetchers.get_all_topics()
+    for topic in topics:
+        story_ids += topic.get_canonical_story_ids()
```
Should I use a set here? (though we don't expect to have a duplicate value, also having duplicate value won't cause any issue.)

DubeySandeep commented on core/domain/opportunity_services.py:
```diff
@@ -384,3 +403,32 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+    story_id_to_story = {}
+    for story in stories:
+        story_id_to_story[story.id] = story
+        exp_ids += story.story_contents.get_all_linked_exp_ids()
```
Ditto*

DubeySandeep commented:
Closed this PR as it needs a different approach to resolve the issue. (I'm going to write one-off job instead)

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into regenerate-opportunity

DubeySandeep committed:
Adds one off job.

DubeySandeep committed:
Added one-off-job files

DubeySandeep committed:
Add proper documentation.

DubeySandeep committed:
Resolved merge conflicts.

DubeySandeep committed:
Resolve lint issues.

DubeySandeep commented on package-lock.json:
```diff
@@ -4486,9 +4486,9 @@
-      "integrity": "sha512-eAZS2sEUMlIeCjBeubdj45dmBHQwPHWyBcT1VSYB7o9x9WRRqKxyUoiXlRjyAwzN7YEzHJlYg0NmzDRWx6GP4Q==",
+      "version": "1.4.0",
+      "resolved": "https://registry.npmjs.org/eslint-utils/-/eslint-utils-1.4.0.tgz",
+      "integrity": "sha512-7ehnzPaP5IIEh1r1tkjuIrxqhNkzUJa9z3R92tLJdZIVdWaczEhr3EbhGtsMrVxi1KeR8qA7Off6SWc5WNQqyQ==",
```
@kevinlee12, Somehow I ended up on this and now I cannot undo this through commit (As pre_commit_hook will undo the change.)

Do you have any suggestion on how I can fix this?

DubeySandeep replied:
I used `--no-verify` flag.

kevinlee12 replied:
Typically, I’d use `git checkout HEAD — package-lock.json` to undo those changes. For this instance, it might have been that the git merge didn’t include this change.

seanlip replied:
/cc @ankita240796  -- I think related to #7452

DubeySandeep replied:
>  For this instance, it might have been that the git merge didn’t include this change.

Yeah, this happened because:

- git pull upstream develop raised merge conflicts 
- Resolved merge conflict
- git commit to merge the changes but it also removed the changes in package-lock.json

> git checkout HEAD — package-lock.json

I think this will also have to go through the `git commit`, right?

kevinlee12 replied:
Yeah, I have a feeling it might be since that will have to go through the pre commit check (which would have undone this change), especially that you have a git conflict.

DubeySandeep committed:
Add more test.

DubeySandeep committed:
Fix space backend test issue

DubeySandeep committed:
Add backend test releated changes.

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into regenerate-opportunity

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,73 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Jobs to run jobs related to opportunity models."""
```
"Jobs to run jobs..."? (Wording is weird.)

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,73 @@
+class ExplorationOpportunitySummaryModelRegenerationJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for regenerating ExplorationOpportunitySummaryModel."""
+    @classmethod
```
Add newline above this one.

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,73 @@
+        exploration_opportunity_summary_list = []
+        for story in stories:
+            for exp_id in story.story_contents.get_all_linked_exp_ids():
+                exploration_opportunity_summary = (
```
exp_opportunity_summary_model (ditto)

DubeySandeep replied:
This is going to be the summary object. (`create_exploration_opportunity_summary_model` is going to be `create_exploration_opportunity_summary`)

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,73 @@
+            exp_ids += story.story_contents.get_all_linked_exp_ids()
+        exp_ids_to_exp = exp_fetchers.get_multiple_explorations_by_id(exp_ids)
+
+        exploration_opportunity_summary_list = []
```
exp_opportunity_summary_models (differentiate between models and domain objects)

DubeySandeep replied:
This is going to be the summary object. (`create_exploration_opportunity_summary_model` is going to be `create_exploration_opportunity_summary`)

seanlip replied:
Ah, that's definitely better. Thanks!

seanlip commented on core/domain/story_fetchers.py:
```diff
@@ -190,6 +190,22 @@ def get_story_summary_by_id(story_id, strict=True):
         return None
 
 
+def get_story_by_ids(story_ids):
```
get_stories_by_ids

DubeySandeep replied:
Done!

seanlip commented on core/domain/story_fetchers.py:
```diff
@@ -190,6 +190,22 @@ def get_story_summary_by_id(story_id, strict=True):
+        story_ids: list(str). List of IDs to get stories for.
+
+    Returns:
+        list(Story). The list of stories corresponding to given ids.
```
Describe the None case too. Ensure it is covered in tests.

DubeySandeep replied:
Done!

seanlip reviewed (changes requested):
Thanks! Left some comments.

sagangwee commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,144 @@
+        (
+            opportunity_jobs_one_off
+            .ExplorationOpportunitySummaryModelRegenerationJob.enqueue(job_id)
+        )
```
Leave ending parenthesis on previous line like the above statement?

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#34-indentation

DubeySandeep replied:
Done!

sagangwee commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,144 @@
+        (
+            opportunity_jobs_one_off
+            .ExplorationOpportunitySummaryModelRegenerationJob.enqueue(job_id)
+        )
```
Leave ending parenthesis on previous line like the above statement?

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#34-indentation

DubeySandeep replied:
Done!

sagangwee commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,144 @@
+            })], 'Changes.')
+
+    def test_regeneration_job(self):
+        """Tests the case where user has no created or edited explorations."""
```
The comment shouldn't be necessary if the test name is sufficiently detailed with the tested behavior and expected outcome.

https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests

DubeySandeep replied:
Done!

sagangwee commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,144 @@
+
+        self.assertTrue(old_creation_time < new_creation_time)
+
+    def test_regeneration_job_for_deleted_topic(self):
```
Include expected outcome in the test name, so something like:

test_regenerationJob_deletedTopic_returnsEmptyOutput

https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests

DubeySandeep replied:
Done!

sagangwee commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,144 @@
+                'new_value': '0'
+            })], 'Changes.')
+
+    def test_regeneration_job(self):
```
Consider splitting up this test. There seems to be a lot being tested here and it's difficult to understand what's going at first glance:

* opportunity_services returns the initial translation opportunity
* the regeneration job returns SUCCESS on a valid process
* ExplorationOpportunitySummaryModel returns the expected number of models
* opportunity_services returns a new opportunity with updated created time after the regeneration job

DubeySandeep replied:
Done!

DubeySandeep committed:
Address review cooments.

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into regenerate-opportunity

DubeySandeep commented:
@sagangwee & @seanlip, I've addressed the review comments PTAL!

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,73 @@
+        exp_ids = []
+        for story in stories:
+            exp_ids += story.story_contents.get_all_linked_exp_ids()
+        exp_ids_to_exp = exp_fetchers.get_multiple_explorations_by_id(exp_ids)
```
We should handle error cases, like e.g. some exp_id has no corresponding exploration, some story_id has no corresponding story, etc. These should be yielded in the logs with identifying information (as specific failure cases) and should also be tested.

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,212 @@
+
+        new_creation_time = all_opportunity_models[0].created_on
+
+        self.assertTrue(old_creation_time < new_creation_time)
```
Can use self.assertLessThan

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,212 @@
+            })], 'Changes.')
+
+
+    def test_regeneration_job_returns_the_initial_opportunity(self):
```
Have only one line between methods in a class.

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off_test.py:
```diff
@@ -0,0 +1,212 @@
+        self.assertEqual(old_opportunities, new_opportunities)
+
+    def test_regeneration_job_returns_correct_output(self):
+        exp_opp_sumary_model_regen_job_class = (
```
sumary --> summary; ditto elsewhere.

DubeySandeep replied:
Done!

seanlip reviewed (changes requested):
Thanks, @DubeySandeep. This looks good! Just one more round of suggestions, and it's good to go (from my perspective).

DubeySandeep committed:
Address review comments.

DubeySandeep committed:
Updated with upstream/develop.

DubeySandeep committed:
Renamed variable

DubeySandeep committed:
Adds unicode modules.

DubeySandeep committed:
Remove extra newline.

DubeySandeep commented:
@seanlip, I've addressed the review comments and added the functionality to generate opportunities for a given topic through the admin page. 
![image](https://user-images.githubusercontent.com/16653571/64135718-0676fe80-ce09-11e9-8f4b-bbda66819b2a.png)

![image](https://user-images.githubusercontent.com/16653571/64135721-0e36a300-ce09-11e9-8ffe-51a8b71a04dc.png)


EDIT:

I've changed the "submit query" button ---> "Regenerate"

DubeySandeep committed:
Changed button name.

seanlip commented:
I'm having some trouble reviewing this, unfortunately. I'm getting this error at the bottom of the "Files Changed" page, and clicking "Retry" does not do anything.

![Screenshot from 2019-09-02 21-36-23](https://user-images.githubusercontent.com/10575562/64144545-d6f4d180-cdc9-11e9-8dfb-1c19f891eae6.png)


@DubeySandeep, are you able to reproduce this? 

EDIT: This seems to have been a transient github error -- please ignore.

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,62 @@
+            yield ('SUCCESS', result['opportunities_count'])
+        else:
+            del result['status']
+            result['topic_id'] = topic_model.id
```
Don't reuse or redefine 'result'. Explicitly construct a new dict inline in line 54 instead to pass to the reduce() method.

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_jobs_one_off.py:
```diff
@@ -0,0 +1,62 @@
+        if key == 'SUCCESS':
+            yield (key, sum(values))
+        else:
+            yield (key + ' (%s)' % len(values), values)
```
Typically we prefer string interpolation, so '%s (%s)' % (..., ...) for the first arg.

DubeySandeep replied:
Done!

seanlip commented on core/domain/opportunity_services.py:
```diff
@@ -386,3 +405,65 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+            opportunities related to the given topic.
+
+    Returns:
+        dict. A dict representing the result of the regeneration process.
```
No, this shouldn't be a dict. Make a domain object for it, or return a tuple with fixed types.

One alternative thing you could perhaps do is to raise an error and let the caller handle it appropriately.

DubeySandeep replied:
Earlier this functionality was only in the jon_one_off file so I had a dict then, now I had to create a method so that I can reuse it in one_off job and in the admin tab, so I just moved the functionality into this method without realizing it's in the service layer. I've added an exception instead. Done!

seanlip commented on core/domain/opportunity_services.py:
```diff
@@ -386,3 +405,65 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+    if len(non_existing_exp_ids) > 0 or len(non_existing_story_ids) > 0:
+        return {
+            'status': 'FAILED',
+            'missing_exp_with_ids': list(non_existing_exp_ids),
```
I don't think you need to programmatize this -- it should be fine to convey this info in an error message string. (You don't actually use the raw values directly anywhere, right?)

DubeySandeep replied:
Done!

seanlip reviewed (changes requested):
Thanks! Just a few notes.

DubeySandeep committed:
Changed service method.

DubeySandeep commented:
@aks681, Can you please review this PR?

@seanlip, I've addressed your review comments PTAL!

seanlip reviewed (approved):
LGTM. Thanks!

seanlip commented:
@aks681 PTAL as codeowner. Thanks!

DubeySandeep commented:
@aks681, Can you please review this PR?

DubeySandeep commented:
@aks681, Can you please review this PR? (This needs to be a merged before the next release cut!)

aks681 commented on core/domain/story_fetchers.py:
```diff
@@ -191,6 +191,23 @@ def get_story_summary_by_id(story_id, strict=True):
+        story_ids: list(str). List of IDs to get stories for.
+
+    Returns:
+        list(Story|None). The list of stories corresponding to given ids.  If an
```
...if a

DubeySandeep replied:
Done!

aks681 commented on core/domain/story_fetchers_test.py:
```diff
@@ -95,6 +95,21 @@ def test_get_story_by_id(self):
         story = story_fetchers.get_story_by_id(self.STORY_ID)
         self.assertEqual(story.to_dict(), expected_story)
 
+    def test_get_stories_by_ids(self):
```
Is this test needed? Since, the test below tests this functionality as well.

DubeySandeep replied:
Yeah, the other one id ofr the None case. (I've changed the test name to make it distinct.)

aks681 reviewed (approved):
lgtm as codeowner! Just 2 comments. Feel free to merge after that.

DubeySandeep committed:
Adds review comments.

DubeySandeep commented:
@nithusha21, This PR needs your approval, can you PTAL?

DubeySandeep commented:
@nithusha21, Can you please review this PR?

nithusha21 reviewed (approved):
Sorry got the delay! LGTM as a codeowner
